### ISSUE TITLE
False positive: when using optional callable function in lambda after asserting callable is not None  

### ISSUE THREAD

**Issue by HiroIshida** [python/mypy#13751]
**Bug Report**
When function takes optional callable, even though callable is asserted that is is not None, mypy says `error: "None" not callable` when used in a lambda function. 

**To Reproduce**
```python
from typing import Callable, Optional, List

def example(fun: Optional[Callable[[int], float]]) -> None:
    assert fun is not None
    sorted([1, 2], key=lambda x: fun(x))
```

**Expected Behavior**
no warning

**Actual Behavior**
show the following warning
```
h-ishida@stonep:~/tmp$ mypy mypy_error.py 
mypy_error.py:6: error: "None" not callable  [misc]
```

**Your Environment**

- Mypy version used: 0.990+dev.9783b46a0062b6acfb8ba58e349e811873fc13f2
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.8.10

**Comment by hauntsaninja** [comment]
https://mypy.readthedocs.io/en/stable/common_issues.html#narrowing-and-inner-functions / see https://github.com/python/mypy/issues/2608
